File: packages/core/test/src/TestEventHub.cxx
/*
 * Copyright (C) 2007 Robotics at Maryland
 * Copyright (C) 2007 Joseph Lisee <jlisee@umd.edu>
 * All rights reserved.
 *
 * Author: Joseph Lisee <jlisee@umd.edu>
 * File:  packages/core/test/src/TestEventHub.cxx
 */


// Library Includes
#include <UnitTest++/UnitTest++.h>
#include <boost/bind.hpp>
#include <boost/thread.hpp>
#include <boost/thread/barrier.hpp>

// Project Includes
#include "core/include/EventHub.h"
#include "core/include/EventPublisher.h"
#include "core/include/EventConnection.h"
#include "core/test/include/Reciever.h"

struct EventHubFixture {
    Reciever recv;
    ram::core::EventHubPtr eventHub;
    ram::core::EventPublisher publisherA;
    ram::core::EventPublisher publisherB;

    EventHubFixture() :
        eventHub(new ram::core::EventHub()),
        publisherA(eventHub),
        publisherB(eventHub)
    {}
};

TEST_FIXTURE(EventHubFixture, subscribe)
{
    Reciever recvB;

    // Subscribe to just events from publisherA
    ram::core::EventConnectionPtr connectionA =
        eventHub->subscribe("Type", &publisherA,
                            boost::bind(&Reciever::handler, &recv, _1));
    
    // Subscribe to just events from publisherB
    ram::core::EventConnectionPtr connectionB =
        eventHub->subscribe("Type", &publisherB,
                            boost::bind(&Reciever::handler, &recvB, _1));
    
    CHECK_EQUAL(0, recv.calls);
    CHECK_EQUAL(0, recvB.calls);    
    
    // Make sure messages from the first publisher gets through
    publisherA.publish("Type", ram::core::EventPtr(new ram::core::Event()));
    CHECK_EQUAL(1, recv.calls);
    CHECK_EQUAL("Type", recv.events[0]->type);
    CHECK_EQUAL(&publisherA, recv.events[0]->sender);
    CHECK_EQUAL(0, recvB.calls);    
    
    // Make sure messages from the second publisher gets through
    publisherB.publish("Type", ram::core::EventPtr(new ram::core::Event()));
    CHECK_EQUAL(1, recvB.calls);
    CHECK_EQUAL("Type", recvB.events[0]->type);
    CHECK_EQUAL(&publisherB, recvB.events[0]->sender);
    CHECK_EQUAL(1, recv.calls);    
    
    // Now for disconnection
    connectionA->disconnect();
    
    // Make sure messages from the first publisher don't get through
    publisherA.publish("Type", ram::core::EventPtr(new ram::core::Event()));
    CHECK_EQUAL(1, recv.calls);
    CHECK_EQUAL(1, recvB.calls);

    publisherB.publish("Type", ram::core::EventPtr(new ram::core::Event()));
    CHECK_EQUAL(1, recv.calls);
    CHECK_EQUAL(2, recvB.calls);

    connectionB->disconnect();
    
    // Make sure messages from the second publisher don't get through
    publisherA.publish("Type", ram::core::EventPtr(new ram::core::Event()));
    publisherB.publish("Type", ram::core::EventPtr(new ram::core::Event()));
    CHECK_EQUAL(1, recv.calls);
    CHECK_EQUAL(2, recvB.calls);
}

TEST_FIXTURE(EventHubFixture, subscribeToType)
{
    ram::core::EventConnectionPtr connection =
        eventHub->subscribeToType("Type",
                                  boost::bind(&Reciever::handler, &recv, _1));
    CHECK_EQUAL(0, recv.calls);    
    
    // Make sure messages from the first publisher gets through
    publisherA.publish("Type", ram::core::EventPtr(new ram::core::Event()));
    CHECK_EQUAL(1, recv.calls);
    CHECK_EQUAL("Type", recv.events[0]->type);
    CHECK_EQUAL(&publisherA, recv.events[0]->sender);
    
    // Make sure messages from the second publisher gets through
    publisherB.publish("Type", ram::core::EventPtr(new ram::core::Event()));
    CHECK_EQUAL(2, recv.calls);
    CHECK_EQUAL("Type", recv.events[1]->type);
    CHECK_EQUAL(&publisherB, recv.events[1]->sender);

    publisherA.publish("AnotherType",
                       ram::core::EventPtr(new ram::core::Event()));
    publisherB.publish("AnotherType",
                       ram::core::EventPtr(new ram::core::Event()));
    CHECK_EQUAL(2, recv.calls);
    
    // Now for disconnection
    connection->disconnect();
    
    // Make sure messages from the first publisher don't get through
    publisherA.publish("Type", ram::core::EventPtr(new ram::core::Event()));
    CHECK_EQUAL(2, recv.calls);
    
    // Make sure messages from the second publisher don't get through
    publisherB.publish("Type", ram::core::EventPtr(new ram::core::Event()));
    CHECK_EQUAL(2, recv.calls);
}

TEST_FIXTURE(EventHubFixture, subscribeToAll)
{
    ram::core::EventConnectionPtr connection =
        eventHub->subscribeToAll(boost::bind(&Reciever::handler, &recv, _1));
    CHECK_EQUAL(0, recv.calls);    
    
    // Make sure messages from the first publisher gets through
    publisherA.publish("Type", ram::core::EventPtr(new ram::core::Event()));
    CHECK_EQUAL(1, recv.calls);
    CHECK_EQUAL("Type", recv.events[0]->type);
    CHECK_EQUAL(&publisherA, recv.events[0]->sender);
    
    // Make sure messages from the second publisher gets through
    publisherB.publish("Type2", ram::core::EventPtr(new ram::core::Event()));
    CHECK_EQUAL(2, recv.calls);
    CHECK_EQUAL("Type2", recv.events[1]->type);
    CHECK_EQUAL(&publisherB, recv.events[1]->sender);

    // Now for disconnection
    connection->disconnect();
    
    // Make sure messages from the first publisher don't get through
    publisherA.publish("Type", ram::core::EventPtr(new ram::core::Event()));
    CHECK_EQUAL(2, recv.calls);
    
    // Make sure messages from the second publisher don't get through
    publisherB.publish("Type", ram::core::EventPtr(new ram::core::Event()));
    CHECK_EQUAL(2, recv.calls);
}

File: wrappers/core/test/src/TestEventHub.py
# Copyright (C) 2007 Maryland Robotics Club
# Copyright (C) 2007 Joseph Lisee <jlisee@umd.edu>
# All rights reserved.
#
# Author: Joseph Lisee <jlisee@umd.edu>
# File: wrapper/core/test/src/TestEventHub.py

# STD Imports
import unittest

# Project Imports
import ext.core as core

class Reciever(object):
    def __init__(self):
        self.etype = None
        self.sender = None
        self.args = None
        self.calls = 0
        self.called = False
    
    def __call__(self, event):
        self.etype = event.type
        self.sender = event.sender
        self.args = event
        self.calls += 1
        self.called = True


class TestEventHub(unittest.TestCase):
    
    def setUp(self):
        self.ehub = core.EventHub()
        self.epubA = core.EventPublisher(self.ehub)
        self.epubB = core.EventPublisher(self.ehub)

    def testSubscribe(self):
        recvA = Reciever()
        recvB = Reciever()
        self.ehub.subscribe("Type", self.epubA, recvA)
        self.ehub.subscribe("Type", self.epubB, recvB)

        self.epubA.publish("Type", core.Event())
        self.assertEquals(1, recvA.calls)
        self.assertEquals(0, recvB.calls)

        self.epubB.publish("Type", core.Event())
        self.assertEquals(1, recvA.calls)
        self.assertEquals(1, recvB.calls)
        
        
    def testSubscribeToType(self):
        recv = Reciever()
        self.ehub.subscribeToType("Type", recv)
        self.assertEquals(0, recv.calls)

        self.epubA.publish("Type", core.Event())
        self.assertEquals(1, recv.calls)

        self.epubB.publish("Type", core.Event())
        self.assertEquals(2, recv.calls)

    def testMakeEventHub(self):
        cfg = { 'name' : 'Test', 'type' : 'EventHub' }
        cfg = core.ConfigNode.fromString(str(cfg))
        ehub = core.SubsystemMaker.newObject(cfg, core.SubsystemList())

        recv = Reciever()
        epub = core.EventPublisher(ehub)
        ehub.subscribeToType("Type", recv)
        
        epub.publish("Type", core.Event())
        self.assertEquals(1, recv.calls)

if __name__ == '__main__':
    unittest.main()
